src/Symfony/Component/DependencyInjection/Compiler/CheckTypeHintsPass.php

```diff
@@ -0,0 +1,165 @@
+                $reflectionMethod->getDeclaringClass()->getName(),
+                $reflectionMethod->getName(),
+                $numberOfRequiredParameters,
+                count($configurationArguments)
+            ));
+        }
+
+        $reflectionParameters = $reflectionMethod->getParameters();
+        $checksCount = min($reflectionMethod->getNumberOfParameters(), count($configurationArguments));
+
+        if ($reflectionMethod->isVariadic()) {
+            --$checksCount;
+        }
+
+        for ($i = 0; $i < $checksCount; ++$i) {
+            if (!$reflectionParameters[$i]->hasType()) {
+                continue;
+            }
+
+            $this->checkTypeHint($configurationArguments[$i], $reflectionParameters[$i]);
+        }
+
+        if ($reflectionMethod->isVariadic() && ($lastParameter = end($reflectionParameters))->hasType()) {
+            $variadicParameters = array_slice($configurationArguments, $lastParameter->getPosition());
+
+            foreach ($variadicParameters as $variadicParameter) {
+                $this->checkTypeHint($variadicParameter, $lastParameter);
+            }
+        }
+    }
+
+    /**
+     * Check type hints compatibility between
+     * a configuration argument and a reflection parameter.
+     *
+     * @param Reference|mixed      $configurationArgument reference of a service or scalar
+     * @param \ReflectionParameter $parameter             the PHP type hinted parameter to compare to
+     *
+     * @throws InvalidArgumentException on type hint incompatibility
+     */
+    private function checkTypeHint($configurationArgument, \ReflectionParameter $parameter)
+    {
+        if ($configurationArgument instanceof Reference) {
+            $argumentDefinition = $this->container->findDefinition((string) $configurationArgument);
+            $class = $this->getClassName($argumentDefinition);
+
+            if (!$this->autoload && !\class_exists($class, false)) {
+                return;
+            }
+
+            if (!\is_a($class, $parameter->getType()->getName(), true)) {
+                throw new InvalidParameterTypeHintException($this->currentId, $class, $parameter);
+            }
+        } else {
+            $checkFunction = 'is_'.$parameter->getType()->getName();
+
+            if (!$parameter->getType()->isBuiltin() || !$checkFunction($configurationArgument)) {
+                throw new InvalidParameterTypeHintException($this->currentId, gettype($configurationArgument), $parameter);
+            }
+        }
+    }
+
+    /**
+     * Get class name from value that can have a factory.
+     *
+     * @param mixed $value
+     *
+     * @return string
+     */
+    private function getClassName($value)
+    {
+        if (is_array($factory = $value->getFactory())) {
+            list($class, $method) = $factory;
+            if ($class instanceof Reference) {
+                $class = $this->container->findDefinition((string) $class)->getClass();
```

Comment:
my mistake I think: here we're looking for the class of a referenced definition - not sure the class of its factory should be used for the validation. just the class of the definition, factory or not, should be more correct. WDYT?

Reply:
I'm not sure about this part as only one test enters inside the `if (is_array($factory = $value->getFactory())) {`, so I'd say factories uses cases are not enough covered